### PR TITLE
Add xView2 Dataset

### PR DESCRIPTION
**Pull request by isaaccorley**
This PR adds the xView2 train/test set used in the [xView2 challenge](https://xview2.org/challenge) and described in ["xBD: A Dataset for Assessing Building Damage from Satellite Imagery", Gupta et al. (2019)](https://arxiv.org/abs/1911.09296)

This allows for a user to load the train/test 1024x1024 imagery and masks as a Vision Dataset, however in the homepage there are additional download links for non chipped imagery and the geojson labels which we should consider adding later as a Geo Dataset. This dataset cannot be automatically downloaded as you must register at the xView2 homepage and the download links expire after a few minutes.

- Added a `draw_semantic_segmentation_mask` method to `torchgeo.datasets.utils` which takes an image tensor, semantic mask tensor (h, w), alpha blend factor, and colormap and overlays the mask onto the image, returning a numpy array which can be used for plotting

TODO:
~~- plot examples~~
- add unit tests

Examples:

image
image

### CODE REVIEW

File: torchgeo/datasets/xview.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+from PIL import Image
+from torch import Tensor
+from torch.utils.data import DataLoader
+from torchvision.transforms import Compose
+
+from ..datasets.utils import dataset_split
+from .geo import VisionDataset
+
+
+class xView2(VisionDataset):
```
Comment (adamjstewart) [inline torchgeo/datasets/xview.py:23]: Classes should use CamelCase: https://www.python.org/dev/peps/pep-0008/#class-names

File: torchgeo/datasets/xview.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    }
+    classes = ["no-damage", "minor-damage", "major-damage", "destroyed"]
+
+    def __init__(
+        self,
+        root: str = "data",
+        split: str = "train",
+        transforms: Optional[Callable[[Dict[str, Tensor]], Dict[str, Tensor]]] = None,
+    ) -> None:
+        """Initialize a new LEVIR-CD+ dataset instance.
```
Comment (adamjstewart) [inline torchgeo/datasets/xview.py:74]: Update name

File: torchgeo/datasets/xview.py (hunk 1 above)
Comment (isaaccorley) [reply torchgeo/datasets/xview.py]: Changed everywhere to XView2

File: torchgeo/datasets/utils.py [hunk 3]
```diff
@@ -32,9 +33,13 @@
     "collate_dict",
     "rasterio_loader",
     "dataset_split",
+    "draw_semantic_segmentation_masks"
 )
 
 
+ColorMap = Union[List[Union[str, Tuple[int, int, int]]], str, Tuple[int, int, int]]
```
Comment (isaaccorley) [inline torchgeo/datasets/utils.py:40]: Taken from https://pytorch.org/vision/master/generated/torchvision.utils.draw_segmentation_masks.html#torchvision.utils.draw_segmentation_masks

**Comment by isaaccorley** [comment]
@adamjstewart sphinx docs builds are failing due to the below errors. Any idea?

```
/home/runner/work/torchgeo/torchgeo/torchgeo/datasets/__init__.py:docstring of torchgeo.datasets.xview.XView2.plot:: WARNING: py:class reference target not found: matplotlib.figure.Figure
/home/runner/work/torchgeo/torchgeo/torchgeo/datasets/__init__.py:docstring of torchgeo.datasets.VisionClassificationDataset:1: WARNING: py:class reference target not found: torchvision.datasets.folder.ImageFolder
```

**Comment by calebrob6** [comment]
@isaaccorley this is from the intersphinx mapping, fix is here https://github.com/microsoft/torchgeo/blob/41502e1b39dfcfff96ee3f6321d24ea35a0f1420/docs/conf.py#L102 (should be merged soon)

**Comment by calebrob6** [comment]
@adamjstewart @isaaccorley I think this is finished now.

Some notes:
- It seems like in many datasets we aren't handing the "checksum" argument properly (i.e. it is only checked in the code path that downloads the dataset).
- This dataset has pre- and post-disaster imagery with pre- and post-disaster building masks/damage labels, so it will not fit nicely with the pre-canned SemanticSegmentationTask.
- We should make it clear that this is the "Challenge training set (~7.8 GB)" and "Challenge test set (~2.6 GB)" data from the xView2 website. I'll do this with any last changes that need to happen.

File: tests/datasets/test_xview2.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import torch
+import torch.nn as nn
+from _pytest.fixtures import SubRequest
+from _pytest.monkeypatch import MonkeyPatch
+
+from torchgeo.datasets import XView2, XView2DataModule
+
+
+class TestXView2:
+    @pytest.fixture(params=["train", "test"])
```
Comment (adamjstewart) [inline tests/datasets/test_xview2.py:19]:
```suggestion
    @pytest.fixture(scope="class", params=["train", "test"])
```

File: torchgeo/datasets/utils.py [hunk 5]
```diff
@@ -431,3 +436,26 @@ def sort_sentinel2_bands(x: str) -> str:
...
+def draw_semantic_segmentation_masks(
+    image: Tensor, mask: Tensor, alpha: float = 0.5, colors: Optional[ColorMap] = None
+) -> np.ndarray:  # type: ignore[type-arg]
+    """Overlay a semantic segmentation mask onto an image.
+
+    Args:
+        image: tensor of shape (3, h, w)
+        mask: tensor of shape (h, w) with pixel values representing the classes
+        alpha: alpha blend factor
+        colors: (Optional) list of RGB int tuples, or color strings e.g. red, #FF00FF
```
Comment (adamjstewart) [inline torchgeo/datasets/utils.py:450]:
```suggestion
        colors: list of RGB int tuples, or color strings e.g. red, #FF00FF
```
It will show up as optional in the docs due to typing

File: torchgeo/datasets/xview.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        if all(exists):
+            return
+
+        # Check if the user requested to download the dataset
+        raise RuntimeError(
+            "Dataset not found in `root` directory, either specify a different"
+            + " `root` directory or manually download the dataset to this directory."
+        )
+
+    @staticmethod
```
Comment (adamjstewart) [inline torchgeo/datasets/xview.py:226]: @calebrob6 I don't think we want these to be static methods in general. I can imagine a dataset like chesapeake or so2sat where the cmap depends on the specific band set a user has chosen when instantiating a dataset. This information can be acquired from `self` if it's a regular class method.

Comment (calebrob6) [reply torchgeo/datasets/xview.py:226]: yeah great point

File: tests/datasets/test_xview2.py (hunk 4 above)
Comment (calebrob6) [reply tests/datasets/test_xview2.py]: This breaks the tests

Comment (calebrob6) [reply tests/datasets/test_xview2.py]: `ScopeMismatch: You tried to access the 'function' scoped fixture 'monkeypatch' with a 'class' scoped request object, involved factories`

**Review by adamjstewart — APPROVED** [review]
